drivers/ethernet/phy/phy_tja1103.c

```diff
@@ -153,140 +151,118 @@ static int phy_tja1103_id(const struct device *dev, uint32_t *phy_id)
-	int rc = 0;
-
-	k_sem_take(&data->sem, K_FOREVER);
+	uint16_t irq;
 
-	/* If Interrupt is configured then the workqueue will not
-	 * update the link state periodically so do it explicitly
-	 */
-	if (cfg->gpio_interrupt.port != NULL) {
-		rc = update_link_state(dev);
+	if (phy_tja1103_c45_read(dev, MDIO_MMD_VENDOR_SPECIFIC1, TJA1103_PHY_FUNC_IRQ_MSTATUS,
+				 &irq) < 0) {
+		return;
 	}
 
-	memcpy(state, &data->state, sizeof(struct phy_link_state));
-
-	k_sem_give(&data->sem);
-
-	return rc;
+	/* Handling Link related Functional IRQs */
+	if (irq & (TJA1103_PHY_FUNC_IRQ_LINK_EVENT | TJA1103_PHY_FUNC_IRQ_LINK_AVAIL)) {
+		/* Ack the assered link related interrupts */
+		phy_tja1103_c45_write(dev, MDIO_MMD_VENDOR_SPECIFIC1, TJA1103_PHY_FUNC_IRQ_ACK,
+				      irq);
+	}
 }
 
-static void invoke_link_cb(const struct device *dev)
+static void phy_tja1103_handle_irq(const struct device *port, struct gpio_callback *cb,
+				   uint32_t pins)
 {
-	struct phy_tja1103_data *const data = dev->data;
-	struct phy_link_state state;
-
-	if (data->cb == NULL) {
-		return;
-	}
+	ARG_UNUSED(pins);
+	ARG_UNUSED(port);
 
-	/* Send callback only on link state change */
-	if (phy_tja1103_get_link_state(dev, &state) != 0) {
-		return;
-	}
+	struct phy_tja1103_data *const data =
+		CONTAINER_OF(cb, struct phy_tja1103_data, phy_tja1103_int_callback);
 
-	data->cb(dev, &state, data->cb_data);
+	/* Trigger workqueue before leaving the ISR */
+	k_work_reschedule(&data->phy_work, K_NO_WAIT);
 }
+#endif
 
-static void monitor_work_handler(struct k_work *work)
+static void phy_work_handler(struct k_work *work)
 {
 	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
 	struct phy_tja1103_data *const data =
-		CONTAINER_OF(dwork, struct phy_tja1103_data, monitor_work);
+		CONTAINER_OF(dwork, struct phy_tja1103_data, phy_work);
 	const struct device *dev = data->dev;
+	struct phy_link_state state = {};
 	int rc;
+	const struct phy_tja1103_config *const cfg = dev->config;
 
-	k_sem_take(&data->sem, K_FOREVER);
-
-	rc = update_link_state(dev);
-
-	k_sem_give(&data->sem);
+	rc = phy_tja1103_get_link_state(dev, &state);
 
-	/* If link state has changed and a callback is set, invoke callback */
-	if (rc == 0) {
-		invoke_link_cb(dev);
+	/* Update link state and trigger callback if changed */
+	if (rc == 0 && (state.speed != data->state.speed || state.is_up != data->state.is_up)) {
+		memcpy(&data->state, &state, sizeof(struct phy_link_state));
+		if (data->cb) {
+			data->cb(dev, &data->state, data->cb_data);
+		}
 	}
 
-	/* Submit delayed work */
-	k_work_reschedule(&data->monitor_work, K_MSEC(CONFIG_PHY_MONITOR_PERIOD));
-}
+	if (cfg->master_slave == 3 && !data->state.is_up &&
+	    k_uptime_delta(&data->start_time) > data->wait_time) {
+		uint16_t val;
 
-static void phy_tja1103_irq_offload_thread(void *p1, void *p2, void *p3)
-{
-	ARG_UNUSED(p2);
-	ARG_UNUSED(p3);
+		data->start_time = k_uptime_get();
+		data->wait_time = 1000 + (sys_rand32_get() % 2001);
```

Comment:
use `sys_timepoint_calc()` and `sys_timepoint_expired()` instead. 

Why is the wait time randomized? is is necessary?